### ISSUE TITLE
`no-useless-spread` it is not safe to remove spread in `[...foo.concat(bar)]`

### ISSUE THREAD

**Issue by magic-akari** [sindresorhus/eslint-plugin-unicorn#2480]
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/2b469bee475a8f3f2767f4669864acdd89654017/test/no-useless-spread.mjs#L259

> [!NOTE]
> The concat() method preserves empty slots if any of the source arrays is [sparse](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Indexed_collections#sparse_arrays).
>
> via [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat#:~:text=The%20concat()%20method%20preserves%20empty%20slots%20if%20any%20of%20the%20source%20arrays%20is%20sparse.)

```js
const foo = new Array(5);
const bar = new Array(5);

const cat = [...foo.concat(bar)];
// now we got a new array with undefined x 10
// but foo.concat(bar) is empty x 10.
// It's not what we want.
```

**Comment by sindresorhus** [comment]
Sparse arrays are an anti-pattern, so we're not going to change the auto-fix, but we should document the issue with sparse arrays, so people can choose to turn off the rule if they want to use sparse arrays.

**Comment by bensaufley** [comment]
So, one way to initialize a _non_-sparse array is `[...new Array(10)]`. This is valid JS, and in this scenario, the spread is not useless: it has a meaningful effect. I definitely understand there are different ways to do that (I use `Array.from({ length: 10 }, () => …)` now), but the above is _valid_ JS that does not result in a sparse array. The fix is unsafe in this scenario. It seems like, at a minimum, the rule should be marked as a potentially unsafe fix.

Interestingly, [eslint's docs](https://eslint.org/docs/latest/rules/no-sparse-arrays) list `new Array(23)` as a "correct" example under `no-sparse-arrays`.

Linters are here to help us write better code, but if you only accept good code to lint in the first place, I feel like that's missing the point of why people use a linter. If the opinion of the linter is

**Comment by fregante** [comment]
> if you only accept good code to lint in the first place, I feel like that's missing the point of why people use a linter

You have a good point here, but I think a lot of unicorn rules build on each other. For example `unicorn/no-new-array` already rewrites it:

```diff
- [...new Array(10)]
+ [...Array.from({length: 10})]
```

Then `unicorn/no-useless-spread` rewrites it again:


```diff
- [...Array.from({length: 10})]
+ Array.from({length: 10})
```

```
$ xo
  index.js:25:1
  ✖  25:1   Unnecessarily cloning an array.                                         unicorn/no-useless-spread
  ✖  25:5   Do not use new Array().                                                 unicorn/no-new-array
```

For your code, this is already being auto-fixed correctly by `xo --fix` (which runs the whole recommended list of unicorn)


```diff
- [...new Array(10)]
+ Array.from({length: 10})
```

I can see the same happening for OP's code:

```diff
- const foo = new Array(5);
- const bar = new Array(5);
+ const foo = Array.from({length: 5});
+ const bar = Array.from({length: 5});

- const cat = [...foo.concat(bar)];
+ const cat = foo.concat(bar);
```

With further advice on the concat, not autofixed:

```
  ✖  28:17  Prefer the spread operator over Array#concat(…).                               unicorn/prefer-spread
```

The result is the same, `cat` has 10 `undefined` items.


In short, the advice to "turn off the rule if they want to use sparse arrays" seems valid here.
